### PR TITLE
Fork the knative/networking Status.Prober

### PR DESCRIPTION
**Pull request by dprotaso**
I've pulled in the knative/networking Prober from https://github.com/knative-extensions/net-gateway-api/pull/672 into this PR.

I've split up the commits for an easier review and better tracking of the changes

tl;dr - 

>### Update the status prober to work with dynamic paths
>
>Because probes on the HTTPRoute now have unique paths
we no longer use a vanilla prober from knative/networking.
>
>The changes make the prober less coupled to Ingress and it requires
consumers to build up a list of backends URLs to probe.
>
>The probe target lister now merges backends and URLs with their
corresponding proxy pod IPs and ports. The alternative
would be to look up HTTPRoutes in the informer cache but theres
no guarantee this is up to date.

### CODE REVIEW

**Comment by dprotaso** [comment]
cc @skonto @ReToCode

**Comment by dprotaso** [comment]
For now I didn't want to change the existing prober in `knative/networking` since I want to first see how this approach works for net-gateway-api.

File: pkg/status/status.go [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+		defer m.mu.Unlock()
+		if state, ok := m.ingressStates[key]; ok {
+			return ProbeState{Version: state.version, Ready: state.pendingCount.Load() == 0}, true
+		}
+		return ProbeState{}, false
+	}()
+
+	return state, ok
+}
+
```
Comment (izabelacg) [inline pkg/status/status.go:196]: Do you mind adding a description to the DoProbes function like IsReady had before?

File: pkg/status/status_test.go [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+	if diff := cmp.Diff(ProbeState{}, state); diff != "" {
+		t.Error("inactive probe should have empty state: ", diff)
+	}
+
+	pstate, err := prober.DoProbes(context.Background(), backends)
+	if err != nil {
+		t.Fatal("IsReady failed:", err)
+	}
+	if pstate.Ready == true {
+		t.Fatal("IsReady() returned true")
```
Comment (izabelacg) [inline pkg/status/status_test.go:171]: I noticed multiple messages in the test file using "IsReady()" text, that could be modified a little. Unless you are already planning to do that in subsequent PRs.

**Comment by izabelacg** [comment]
Thanks for the multiple commits, it was easier to digest :). I added just minor comments.

File: pkg/status/status.go [hunk 3]
```diff
@@ -174,7 +194,7 @@ func (m *Prober) IsReady(ctx context.Context, ing *v1alpha1.Ingress) (bool, erro
 		m.mu.Lock()
 		defer m.mu.Unlock()
 		if state, ok := m.ingressStates[ingressKey]; ok {
-			if state.hash == hash {
+			if state.version == hash {
```
Comment (ReToCode) [inline pkg/status/status.go]: Rename the hash on the right as well to make it consistent?

File: pkg/status/status.go [hunk 4]
```diff
@@ -155,6 +160,21 @@ func NewProber(
 	}
 }
 
+func (m *Prober) IsProbeActive(ing *v1alpha1.Ingress) (ProbeState, bool) {
+	ingressKey := types.NamespacedName{Namespace: ing.Namespace, Name: ing.Name}
+
+	state, ok := func() (ProbeState, bool) {
```
Comment (ReToCode) [inline pkg/status/status.go:185]: Is there a reason (that I'm not seeing) that this is wrapped in a immediately executed inline function?

File: pkg/status/status.go [hunk 5]
```diff
@@ -175,52 +197,60 @@ func (m *Prober) IsProbeActive(ing *v1alpha1.Ingress) (ProbeState, bool) {
...
-	hash := fmt.Sprintf("%x", bytes)
-
-	if ready, ok := func() (bool, bool) {
+func (m *Prober) DoProbes(ctx context.Context, backends Backends) (ProbeState, error) {
+	if state, ok := func() (ProbeState, bool) {
 		m.mu.Lock()
 		defer m.mu.Unlock()
-		if state, ok := m.ingressStates[ingressKey]; ok {
-			if state.version == hash {
+		if state, ok := m.ingressStates[backends.Key]; ok {
```
Comment (ReToCode) [inline pkg/status/status.go:201]: maybe name `state` `ingState` to make it more clear compared to e.g. `probeState`?

File: pkg/status/status_test.go (hunk 2 above)
Comment (ReToCode) [reply pkg/status/status_test.go:171]: +1, also the error message would be misleading (was before as well).

File: pkg/status/status_test.go [hunk 6]
```diff
@@ -248,22 +249,30 @@ func TestProbeLifecycle(t *testing.T) {
 		<-cancelled
 	}()
 
+	backends := Backends{
```
Comment (ReToCode) [inline pkg/status/status_test.go:299]: maybe extract this to a helper?

File: pkg/status/status_test.go [hunk 7]
```diff
@@ -579,113 +621,12 @@ func TestPartialPodCancellation(t *testing.T) {
 	}
 }
 
-func TestCancelIngressProbing(t *testing.T) {
```
Comment (ReToCode) [inline pkg/status/status_test.go]: Can you elaborate why this is no longer needed?

File: pkg/status/status.go (hunk 3 above)
Comment (dprotaso) [reply pkg/status/status.go]: Done in a later commit

File: pkg/status/status.go (hunk 4 above)
Comment (dprotaso) [reply pkg/status/status.go:185]: No I can dropped this - I copied it from the earlier IsReady that did extra work so it didn't want to hold the lock

File: pkg/status/status_test.go (hunk 7 above)
Comment (dprotaso) [reply pkg/status/status_test.go]: Was dropped unintentionally

File: pkg/status/status.go (hunk 5 above)
Comment (dprotaso) [reply pkg/status/status.go]: done

File: pkg/status/status_test.go (hunk 2 above)
Comment (dprotaso) [reply pkg/status/status_test.go]: done

File: pkg/status/status.go (hunk 1 above)
Comment (dprotaso) [reply pkg/status/status.go:193]: done

File: pkg/status/status_test.go (hunk 6 above)
Comment (dprotaso) [reply pkg/status/status_test.go:299]: Yeah I debated and I thought the host, hash, visibillity vary enough and the struct was small enough that I didn't feel it necessary.

**Comment by dprotaso** [comment]
Thanks for the reviews @izabelacg and @ReToCode 

I've addressed the feedback please take another look

File: pkg/status/status_test.go [hunk 8]
```diff
@@ -0,0 +1,900 @@
...
+		Version: "some-hash",
+		URLs: map[v1alpha1.IngressVisibility]URLSet{
+			v1alpha1.IngressVisibilityExternalIP: sets.New(
+				url.URL{Scheme: "http", Host: "foo.bar.com"},
+			),
+		},
+	}
+	state, err := prober.DoProbes(ctx, backends)
+	if err != nil {
+		t.Fatal("IsReady failed:", err)
```
Comment (izabelacg) [inline pkg/status/status_test.go:726]: I think you missed one "IsReady"

**Pull request merged**